Add conversion of German Umlauts to “ae”, “oe”, “ue” in Home Assistant Entity Names

mschabhuettl opened the pull request:
This pull request addresses the issue with entity name generation in Home Assistant. The code changes ensure accurate conversion of characters with diacritical marks, specifically focusing on German Umlauts (ä, ö, ü).

Reference: [Conversion of German Umlauts to “ae”, “oe”, “ue” in Home Assistant Entity Names](https://community.home-assistant.io/t/conversion-of-german-umlauts-to-ae-oe-ue-in-home-assistant-entity-names/591327)

The modification to the slugify() function includes the necessary replacements of GERMAN_TRANSLATIONS. With these replacements incorporated, the conversion of German Umlauts will now be handled correctly, resulting in accurate and consistent entity names for German-speaking users.

While I have not personally tested the code changes, I have reviewed and validated the modifications, and they are expected to effectively resolve the problem.

I kindly request the Home Assistant development team to review and merge this pull request. I am excited to contribute to the ongoing enhancement of Home Assistant and help create a better experience for users worldwide.

Thank you for considering this pull request.

Best regards,
Matthias Schabhüttl

mschabhuettl committed:
add german umlaut support

by using replacements option from slugify, german umlaut conversion should be fixed (ä>ae, ö>oe, ü>ue, etc.).

mschabhuettl committed:
fix slugify

mschabhuettl committed:
Merge branch 'dev' into patch-1

mschabhuettl committed:
Merge branch 'dev' into patch-1

mschabhuettl committed:
update name

mschabhuettl commented:
Should work now, but needs to be tested. Original PRE_TRANSLATION does not include ß-Character. @epenet

mschabhuettl committed:
Merge branch 'dev' into patch-1

mschabhuettl committed:
Merge branch 'dev' into patch-1

mschabhuettl committed:
Merge branch 'dev' into patch-1

mschabhuettl committed:
Merge branch 'dev' into patch-1

mschabhuettl committed:
Merge branch 'dev' into patch-1

oliv3r commented:
Not sure what the HA policy is, but I guess a general remark would be to rebase/cleanup the branches commit history, it's a little messy atm.

oliv3r commented on homeassistant/util/__init__.py:
```diff
@@ -44,7 +44,13 @@ def slugify(text: str | None, *, separator: str = "_") -> str:
+    ['ü', 'ue'],
+    ['ß', 'ss'],
+    ]
+    slug = unicode_slug.slugify(text, separator=separator, replacements=GERMAN_TRANSLATIONS)
```
While these are very valid in german, there's multiple languages that suffer from the same? Is this something that could be made more generic, and not specifcally only for GERMAN? if anything, does the variable need to be named GERMAN, so that in the future this could be re-used for other languages?

mschabhuettl replied:
Hmm... The Variable could be named UMLAUT_TRANSLATIONS or just TRANSLATIONS. Maybe even PRE_TRANSLATIONS like the original module in slugify.

mschabhuettl replied:
It has been pointed out that some Icelandic and Scandinavian languages also use ö or similar characters. However, as I am not a linguist or an expert in this field, and I do not speak these languages (only English and German), I cannot provide specific advice in this regard.

oliv3r replied:
I would think that's fine, but at least we'd leave the door open for icelandic and scandiavian (etc) speakers to add their translations too :)

mschabhuettl replied:
Fixed variable name in commit 511edbfa78e2c0832c960358bd9dc40273820f86. Sorry, can't rebase right now as I'm on a webbrowser now. Thought I shoud keep the branch up-to-date with current upstream, so i merged changes from dev.

mschabhuettl committed:
changed to PRE_TRANSLATIONS

changed GERMAN_TRANSLATIONS variable name to PRE_TRANSLATIONS to keep it generic.

emontnemery commented:
> Not sure what the HA policy is, but I guess a general remark would be to rebase/cleanup the branches commit history, it's a little messy atm; 10 commits for a 6? line change :)

If a PR has already been reviewed, we don't want history to be rewritten by squashing commits in a rebase. Several commits are normally fine if it's within reason.

emontnemery reviewed (changes requested):
While I understand the rationale for more readable German strings, changing the slugify algorithm would be a huge breaking change.
Also, German is not the only language using the umlaut, so while this would be a readability improvement for German it may not be an improvement for other languages.

If we go ahead with this, tests need to be updated, for example like this:
```diff
diff --git a/tests/util/test_init.py b/tests/util/test_init.py
index 3ead4cf41c8..6f3d9222f20 100644
--- a/tests/util/test_init.py
+++ b/tests/util/test_init.py
@@ -60,6 +60,12 @@ def test_slugify() -> None:
     assert util.slugify("") == ""
     assert util.slugify(None) == ""

+    # Custom German replacements
+    assert util.slugify("Käfer") == "kaefer"
+    assert util.slugify("Döner") == "doener"
+    assert util.slugify("müde") == "muede"
+    assert util.slugify("Straße") == "strasse"
+
```

emontnemery commented:
@mschabhuettl I just want to highlight that while changing the slugify algorithm used all over Home Assistant core to improve readability of entity ids is not a good solution, the problem of imperfect (in case of German) or outright unreadable entity ids (in case of languages not using the Latin alphabet) is well understood.

A discussion here <https://github.com/home-assistant/architecture/discussions> highlighting the problem and exploring different solutions would be welcome. If we find a way forward there, the next step would be another core PR implementing the agreed upon changes.
